go/vt/topo/server.go

```diff
@@ -294,6 +294,28 @@ func (ts *Server) ConnForCell(ctx context.Context, cell string) (Conn, error) {
 	return conn, nil
 }
 
+// CellToRegionMapper function is a wrapper around topo.Server#GetRegionByCell with caching and error handling
```

Comment:
I don't see the benefit of using this pattern, vs just having a global map and a function to get the value from the map directly. Also, the global function backed by a global map would be easy to unit test. And then you wouldn't need to pass it everywhere. Overall, the number of lines of code for this change would go down dramatically.

Also, this needs to be thread-safe, so please protect the map with a mutex.

Reply:
Why do you need to collect anything upfront? You start with an empty map and a mutex:

var cellToRegionMap map[string]string
var cellToRegionMapMu sync.Mutex

and export a function that does the usual Mutex lock, get from the cache, if not there get from topo service, and populate the cache.

The population only happens the first time you call the function, there is nothing upfront needed... You can also use a RW Mutex and do the lookup holding just the Read lock.

It's not different from what you're doing, except it's a global cache?

Reply:
👍